src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5PoNCreator.java

```diff
@@ -0,0 +1,712 @@
+            final double[] rowCounts = counts.getRow(i);
+            int zeroCount = 0;
+            int nonZeroCount = 0;
+            for (int j = 0; j <  rowCounts.length; j++) {
+                if (rowCounts[j] != 0.0) {
+                    nonZeroValues[nonZeroCount++] = rowCounts[j];
+                } else {
+                    zeroIndexes[zeroCount++] = j;
+                }
+            }
+            totalCounts += rowCounts.length;
+            if (zeroCount > 0) { // We only have to do anything if if there is any zeros in the row.
+                final double nonZeroMedian = medianCalculator.evaluate(nonZeroValues, 0, nonZeroCount);
+                for (int j = 0; j < zeroCount; j++) {
+                    counts.setEntry(i, zeroIndexes[j], nonZeroMedian);
+                }
+                totalZeroCounts += zeroCount;
+            }
+        }
+        if (totalZeroCounts > 0) {
+            logger.info(String.format("Some 0.0 counts, %d of %d (%.2f%%), where imputed to their enclosing target non-zero median", totalZeroCounts, totalZeroCounts, 100.0 * (totalZeroCounts / totalCounts)));
+        } else {
+            logger.info("No count is 0.0 thus no count needed to be imputed.");
+        }
+    }
+    /**
+     * Truncates the extreme count values in the input read-count collection.
+     * Values are forced to be bound by the percentile indicated with the input {@code percentile} which must be
+     * in the range [0 .. 50.0]. Values under that percentile and the complementary (1 - percentile) are set to the
+     * corresponding threshold value.
+     *
+     * <p>The imputation is done in-place, thus the input matrix is well be modified as a result of this call.</p>
+     *
+     * @param readCounts the input and output read-count matrix.
+     */
+    @VisibleForTesting
+    static void truncateExtremeCounts(final ReadCountCollection readCounts, final double percentile, final Logger logger) {
+
+        final RealMatrix counts = readCounts.counts();
+        final int targetCount = counts.getRowDimension();
+        final int columnCount = counts.getColumnDimension();
+        final double[] values = new double[targetCount * columnCount];
+        for (int i = 0; i < targetCount; i++) {
+            final double[] rowCounts = counts.getRow(i);
+            System.arraycopy(rowCounts, 0, values, i * columnCount, columnCount);
+        }
+        final Percentile bottomPercentileEvaluator = new Percentile(percentile);
+        final Percentile topPercentileEvaluator = new Percentile(100.0 - percentile);
+        final double bottomPercentileThreshold = bottomPercentileEvaluator.evaluate(values);
+        final double topPercentileThreshold = topPercentileEvaluator.evaluate(values);
+        long totalCounts = 0;
+        long bottomTruncatedCounts = 0;
+        long topTruncatedCounts = 0;
+
+        for (int i = 0; i < targetCount; i++) {
```

Comment:
I'm a little worried about using visitors, since there could be a performance hit when preserving the log messages.  I did the conversion above, but have to add a second pass through the data just to calculate the number of values that will be imputed. 

Here I would have to do two additional passthroughs.  This may not be a problem, because the optimizer can probably figure out what is going on.  But if it can't this would be a performance hit.

Tell me if you disagree.

No action.